storm-core/src/jvm/org/apache/storm/daemon/supervisor/ReadClusterState.java

```diff
@@ -0,0 +1,318 @@
+    
+    protected Map<Integer, LocalAssignment> readMyExecutors(String stormId, String assignmentId, Assignment assignment) {
+        Map<Integer, LocalAssignment> portTasks = new HashMap<>();
+        Map<Long, WorkerResources> slotsResources = new HashMap<>();
+        Map<NodeInfo, WorkerResources> nodeInfoWorkerResourcesMap = assignment.get_worker_resources();
+        if (nodeInfoWorkerResourcesMap != null) {
+            for (Map.Entry<NodeInfo, WorkerResources> entry : nodeInfoWorkerResourcesMap.entrySet()) {
+                if (entry.getKey().get_node().equals(assignmentId)) {
+                    Set<Long> ports = entry.getKey().get_port();
+                    for (Long port : ports) {
+                        slotsResources.put(port, entry.getValue());
+                    }
+                }
+            }
+        }
+        Map<List<Long>, NodeInfo> executorNodePort = assignment.get_executor_node_port();
+        if (executorNodePort != null) {
+            for (Map.Entry<List<Long>, NodeInfo> entry : executorNodePort.entrySet()) {
+                if (entry.getValue().get_node().equals(assignmentId)) {
+                    for (Long port : entry.getValue().get_port()) {
+                        LocalAssignment localAssignment = portTasks.get(port.intValue());
+                        if (localAssignment == null) {
+                            List<ExecutorInfo> executors = new ArrayList<ExecutorInfo>();
+                            localAssignment = new LocalAssignment(stormId, executors);
+                            if (slotsResources.containsKey(port)) {
+                                localAssignment.set_resources(slotsResources.get(port));
+                            }
+                            portTasks.put(port.intValue(), localAssignment);
+                        }
+                        List<ExecutorInfo> executorInfoList = localAssignment.get_executors();
+                        executorInfoList.add(new ExecutorInfo(entry.getKey().get(0).intValue(), entry.getKey().get(entry.getKey().size() - 1).intValue()));
+                    }
+                }
+            }
+        }
+        return portTasks;
+    }
+
+    public synchronized void shutdownAllWorkers() {
+        for (Slot slot: slots.values()) {
+            try {
+                slot.setNewAssignment(null);
+            } catch (Exception e) {
+                LOG.error("Error trying to shutdown workers in {}", slot, e);
+            }
+        }
+
+        for (Slot slot: slots.values()) {
+            try {
+                int count = 0;
+                while (slot.getMachineState() != MachineState.EMPTY) {
+                    if (count > 10) {
+                        LOG.warn("DONE waiting for {} to finish {}", slot, slot.getMachineState());
+                        break;
+                    }
+                    if (Time.isSimulating()) {
+                        Time.advanceTime(1000);
+                        Thread.sleep(100);
```

Comment:
We are simulating, so in general we don't actually wait as long in simulated time, to speed up the tests.